flojoy/flojoy_node_venv.py

```diff
@@ -0,0 +1,142 @@
+        multiprocessing.set_executable(self.original_executable_path)
+
+class SwapSysPath:
+    """Temporarily swap the sys.path of the child process with the sys.path of the parent process."""
+    def __init__(self, venv_executable):
+        self.new_path = _get_venv_syspath(venv_executable)
+        self.old_path = None
+
+    def __enter__(self):
+        self.old_path = sys.path
+        sys.path = self.new_path
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        sys.path = self.old_path
+
+
+def _install_pip_dependencies(
+        venv_executable: os.PathLike,
+        pip_dependencies: tuple[str],
+        verbose: bool = False
+    ):
+    """Install pip dependencies into the virtual environment."""
+    command = [venv_executable, "-m", "pip", "install"]
+    if(not verbose):
+        command += ["-q"]  
+    command += list(pip_dependencies)
+    subprocess.check_call(command)
+
+def _get_venv_syspath(venv_executable: os.PathLike) -> list[str]:
+    """Get the sys.path of the virtual environment."""
+    command = [venv_executable, "-c", "import sys\nprint(sys.path)"]
+    cmd_output = subprocess.run(command, check=True, capture_output=True, text=True)
+    return eval(cmd_output.stdout)
+
+class PickleableFunctionWithPipeIO:
+    """A wrapper for a function that can be pickled and executed in a child process."""
+    def __init__(self, func, child_conn, venv_executable):
+        self._func_serialized = cloudpickle.dumps(func)
+        self._child_conn = child_conn
+        self._venv_executable = venv_executable
+
+    def __call__(self, *args, **kwargs):
+        fn = cloudpickle.loads(self._func_serialized)
+        args = [cloudpickle.loads(arg) for arg in args]
+        kwargs = {key: cloudpickle.loads(value) for key, value in kwargs.items()}
+        with SwapSysPath(venv_executable=self._venv_executable):
+            result = fn(*args, **kwargs)
+        self._child_conn.send(result)
+
+def run_in_venv(pip_dependencies=[]):
+    # Pre-pend flojoy and cloudpickle as mandatory pip dependencies
+    pip_dependencies = sorted(["flojoy", "cloudpickle"] + pip_dependencies)
+    # Get the system's temporary directory path
+    temp_dirpath = tempfile.gettempdir()
+    # Generate a path-safe hash of the pip dependencies
+    # this prevents the duplication of virtual environments
+    pip_dependencies_hash = hashlib.sha256("".join(pip_dependencies).encode()).hexdigest()
+    venv_path = os.path.join(temp_dirpath, f"flojoy_node_venv_{pip_dependencies_hash}")
+    venv_executable = os.path.join(venv_path, "bin", "python")
+    # Create the node_env virtual environment if it does not exist
+    if not os.path.exists(venv_path):
+        venv.create(venv_path, with_pip=True)
+    # Install the pip dependencies into the virtual environment
+    if pip_dependencies:
+        _install_pip_dependencies(
+            venv_executable=venv_executable,
+            pip_dependencies=tuple(pip_dependencies)
+        )
```

Comment:
should this block be inside the condition `if not os.path.exists(venv_path):`?

Reply:
Yes